{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import ConfusionMatrixDisplay\n",
    "import numpy as np\n",
    "## need biopython to run: https://anaconda.org/anaconda/biopython\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.Seq import Seq\n",
    "## need OligoArrayAux to run: http://www.unafold.org/Dinamelt/software/OligoArrayAux.php\n",
    "import oligo_melting as OligoMelt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Featurization\n",
    "### Functions\n",
    "can definitely be improved upon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def letter_at_position(string, letter, position):\n",
    "    ## takes in a string, letter (both as strings), and position as an integer\n",
    "    ## returns boolean whether the given letter is at the given position in the string\n",
    "    return string[position:position+1].__eq__(letter)\n",
    "\n",
    "def tuple_at_position(string, letters, position):\n",
    "    ## takes in a string, 2 letters (as a single string) and position as an integer\n",
    "    ## returns boolean whether the given letters are at the given position in the string\n",
    "    return string[position:position+2].__eq__(letters)\n",
    "\n",
    "def triple_at_position(string, letters, position):\n",
    "    ## takes in a string, 3 letters (as a single string) and position as an integer\n",
    "    ## returns boolean whether the given letters are at the given position in the string\n",
    "    return string[position:position+3].__eq__(letters)\n",
    "\n",
    "def find_PAM(sequence, PAM):\n",
    "    ## takes in sequence and PAM as strings representing genetic target sequence and the promotospacer-adjacent motif\n",
    "    ## of the gene editing complex\n",
    "    ## returns the position as an int or -1 if not found\n",
    "    match = re.search(PAM, sequence)\n",
    "    return match.start()\n",
    "\n",
    "def gc_count(string):\n",
    "    ## takes in a string\n",
    "    ## returns an integer representing the number of G or C letters in that string\n",
    "    count = 0\n",
    "    for letter in string:\n",
    "        if letter.__eq__('G') | letter.__eq__('C'):\n",
    "            count=count+1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Doench et al 2014 DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Table Legends', 'Supp Table 1', 'Supp Table 2', 'Supp Table 3', 'Supp Table 4', 'Supp Table 5', 'Supp Table 6', 'Supp Table 7', 'Supp Table 8', 'Supp Table 9', 'Supp Table 10']\n"
     ]
    }
   ],
   "source": [
    "df_doench=pd.ExcelFile(\"Doench 2014 Supplemental Tables.xlsx\")\n",
    "print(df_doench.sheet_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>prob</th>\n",
       "      <th>oddsratio</th>\n",
       "      <th>quantity</th>\n",
       "      <th>odds</th>\n",
       "      <th>stdev</th>\n",
       "      <th>delta_devs</th>\n",
       "      <th>p-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A01</td>\n",
       "      <td>0.187970</td>\n",
       "      <td>0.996638</td>\n",
       "      <td>399</td>\n",
       "      <td>0.231481</td>\n",
       "      <td>0.019579</td>\n",
       "      <td>-0.026282</td>\n",
       "      <td>0.979032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A02</td>\n",
       "      <td>0.207182</td>\n",
       "      <td>1.125124</td>\n",
       "      <td>362</td>\n",
       "      <td>0.261324</td>\n",
       "      <td>0.020556</td>\n",
       "      <td>0.909617</td>\n",
       "      <td>0.363025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A03</td>\n",
       "      <td>0.164835</td>\n",
       "      <td>0.849765</td>\n",
       "      <td>364</td>\n",
       "      <td>0.197368</td>\n",
       "      <td>0.020499</td>\n",
       "      <td>-1.153676</td>\n",
       "      <td>0.248633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A04</td>\n",
       "      <td>0.209877</td>\n",
       "      <td>1.143642</td>\n",
       "      <td>405</td>\n",
       "      <td>0.265625</td>\n",
       "      <td>0.019434</td>\n",
       "      <td>1.100761</td>\n",
       "      <td>0.271001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A05</td>\n",
       "      <td>0.172324</td>\n",
       "      <td>0.896408</td>\n",
       "      <td>383</td>\n",
       "      <td>0.208202</td>\n",
       "      <td>0.019984</td>\n",
       "      <td>-0.808677</td>\n",
       "      <td>0.418701</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Feature      prob  oddsratio  quantity      odds     stdev  delta_devs  \\\n",
       "0     A01  0.187970   0.996638       399  0.231481  0.019579   -0.026282   \n",
       "1     A02  0.207182   1.125124       362  0.261324  0.020556    0.909617   \n",
       "2     A03  0.164835   0.849765       364  0.197368  0.020499   -1.153676   \n",
       "3     A04  0.209877   1.143642       405  0.265625  0.019434    1.100761   \n",
       "4     A05  0.172324   0.896408       383  0.208202  0.019984   -0.808677   \n",
       "\n",
       "    p-value  \n",
       "0  0.979032  \n",
       "1  0.363025  \n",
       "2  0.248633  \n",
       "3  0.271001  \n",
       "4  0.418701  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_table_features=pd.read_excel(\"Doench 2014 Supplemental Tables.xlsx\", sheet_name=\"Supp Table 8\")\n",
    "df_table_features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Spacer Sequence</th>\n",
       "      <th>Expanded Sequence</th>\n",
       "      <th>Spacer_ID</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Essential Gene Set</th>\n",
       "      <th>Expression Value (CCLE)</th>\n",
       "      <th>Early Time Point</th>\n",
       "      <th>DMSO Day 14 Rep 1</th>\n",
       "      <th>DMSO Day 14 Rep 2</th>\n",
       "      <th>ETP - DMSO</th>\n",
       "      <th>Gene Avg</th>\n",
       "      <th>Normalized sgRNA Activity</th>\n",
       "      <th>sgRNA Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTAATGGCTTCCTCGTGAGT</td>\n",
       "      <td>GGCCGTAATGGCTTCCTCGTGAGTTGGTCC</td>\n",
       "      <td>s_40338</td>\n",
       "      <td>PARS2</td>\n",
       "      <td>aminoacyl</td>\n",
       "      <td>7.013512</td>\n",
       "      <td>3.623884</td>\n",
       "      <td>3.727199</td>\n",
       "      <td>2.739853</td>\n",
       "      <td>0.390358</td>\n",
       "      <td>1.173702</td>\n",
       "      <td>-0.783345</td>\n",
       "      <td>0.325738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTCGTGCGAGTGATAGACC</td>\n",
       "      <td>AGAAGCTCGTGCGAGTGATAGACCAGGAGA</td>\n",
       "      <td>s_40337</td>\n",
       "      <td>PARS2</td>\n",
       "      <td>aminoacyl</td>\n",
       "      <td>7.013512</td>\n",
       "      <td>3.432073</td>\n",
       "      <td>0.942266</td>\n",
       "      <td>2.950215</td>\n",
       "      <td>1.485832</td>\n",
       "      <td>1.173702</td>\n",
       "      <td>0.312129</td>\n",
       "      <td>0.162559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CACGGACGGTATATGGCAGG</td>\n",
       "      <td>ATGGCACGGACGGTATATGGCAGGAGGTGG</td>\n",
       "      <td>s_40336</td>\n",
       "      <td>PARS2</td>\n",
       "      <td>aminoacyl</td>\n",
       "      <td>7.013512</td>\n",
       "      <td>4.679738</td>\n",
       "      <td>3.576145</td>\n",
       "      <td>2.493495</td>\n",
       "      <td>1.644918</td>\n",
       "      <td>1.173702</td>\n",
       "      <td>0.471215</td>\n",
       "      <td>0.035653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACTTCTGCGAGACACGGAC</td>\n",
       "      <td>TGGCTACTTCTGCGAGACACGGACAGGTCG</td>\n",
       "      <td>s_32891</td>\n",
       "      <td>MARS2</td>\n",
       "      <td>aminoacyl</td>\n",
       "      <td>7.065261</td>\n",
       "      <td>2.050553</td>\n",
       "      <td>1.291687</td>\n",
       "      <td>1.034652</td>\n",
       "      <td>0.887383</td>\n",
       "      <td>1.086888</td>\n",
       "      <td>-0.199505</td>\n",
       "      <td>0.311042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGGCCCACGCTACTACAGTT</td>\n",
       "      <td>ACTTCGGCCCACGCTACTACAGTTCGGGCT</td>\n",
       "      <td>s_32889</td>\n",
       "      <td>MARS2</td>\n",
       "      <td>aminoacyl</td>\n",
       "      <td>7.065261</td>\n",
       "      <td>5.252744</td>\n",
       "      <td>4.041662</td>\n",
       "      <td>4.907459</td>\n",
       "      <td>0.778184</td>\n",
       "      <td>1.086888</td>\n",
       "      <td>-0.308705</td>\n",
       "      <td>0.199951</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Spacer Sequence               Expanded Sequence Spacer_ID Gene Symbol  \\\n",
       "0  GTAATGGCTTCCTCGTGAGT  GGCCGTAATGGCTTCCTCGTGAGTTGGTCC   s_40338       PARS2   \n",
       "1  GCTCGTGCGAGTGATAGACC  AGAAGCTCGTGCGAGTGATAGACCAGGAGA   s_40337       PARS2   \n",
       "2  CACGGACGGTATATGGCAGG  ATGGCACGGACGGTATATGGCAGGAGGTGG   s_40336       PARS2   \n",
       "3  TACTTCTGCGAGACACGGAC  TGGCTACTTCTGCGAGACACGGACAGGTCG   s_32891       MARS2   \n",
       "4  CGGCCCACGCTACTACAGTT  ACTTCGGCCCACGCTACTACAGTTCGGGCT   s_32889       MARS2   \n",
       "\n",
       "  Essential Gene Set  Expression Value (CCLE)  Early Time Point  \\\n",
       "0          aminoacyl                 7.013512          3.623884   \n",
       "1          aminoacyl                 7.013512          3.432073   \n",
       "2          aminoacyl                 7.013512          4.679738   \n",
       "3          aminoacyl                 7.065261          2.050553   \n",
       "4          aminoacyl                 7.065261          5.252744   \n",
       "\n",
       "   DMSO Day 14 Rep 1  DMSO Day 14 Rep 2  ETP - DMSO  Gene Avg  \\\n",
       "0           3.727199           2.739853    0.390358  1.173702   \n",
       "1           0.942266           2.950215    1.485832  1.173702   \n",
       "2           3.576145           2.493495    1.644918  1.173702   \n",
       "3           1.291687           1.034652    0.887383  1.086888   \n",
       "4           4.041662           4.907459    0.778184  1.086888   \n",
       "\n",
       "   Normalized sgRNA Activity  sgRNA Score  \n",
       "0                  -0.783345     0.325738  \n",
       "1                   0.312129     0.162559  \n",
       "2                   0.471215     0.035653  \n",
       "3                  -0.199505     0.311042  \n",
       "4                  -0.308705     0.199951  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_doench=pd.read_excel(\"Doench 2014 Supplemental Tables.xlsx\", sheet_name=\"Supp Table 10\")\n",
    "df_doench.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fusi et al 2015"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>30mer</th>\n",
       "      <th>Target gene</th>\n",
       "      <th>Percent Peptide</th>\n",
       "      <th>Amino Acid Cut position</th>\n",
       "      <th>score_drug_gene_rank</th>\n",
       "      <th>score_drug_gene_threshold</th>\n",
       "      <th>drug</th>\n",
       "      <th>predictions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>CAGAAAAAAAAACACTGCAACAAGAGGGTA</td>\n",
       "      <td>CD5</td>\n",
       "      <td>72.87</td>\n",
       "      <td>360.0</td>\n",
       "      <td>0.083682</td>\n",
       "      <td>0</td>\n",
       "      <td>nodrug</td>\n",
       "      <td>0.544412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TTTTAAAAAACCTACCGTAAACTCGGGTCA</td>\n",
       "      <td>NF1</td>\n",
       "      <td>65.80</td>\n",
       "      <td>1868.0</td>\n",
       "      <td>0.868207</td>\n",
       "      <td>1</td>\n",
       "      <td>PLX_2uM</td>\n",
       "      <td>0.617512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TCAGAAAAAGCAGCGTCAGTGGATTGGCCC</td>\n",
       "      <td>CD5</td>\n",
       "      <td>84.21</td>\n",
       "      <td>416.0</td>\n",
       "      <td>0.184100</td>\n",
       "      <td>0</td>\n",
       "      <td>nodrug</td>\n",
       "      <td>0.476232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>AATAAAAAATAGGATTCCCAGCTTTGGAAG</td>\n",
       "      <td>NF1</td>\n",
       "      <td>56.39</td>\n",
       "      <td>1601.0</td>\n",
       "      <td>0.432065</td>\n",
       "      <td>0</td>\n",
       "      <td>PLX_2uM</td>\n",
       "      <td>0.459882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>GATGAAAAATATGTAAACAGCATTTGGGAC</td>\n",
       "      <td>CUL3</td>\n",
       "      <td>4.30</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.149351</td>\n",
       "      <td>0</td>\n",
       "      <td>PLX_2uM</td>\n",
       "      <td>0.290841</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                           30mer Target gene  Percent Peptide  \\\n",
       "0           0  CAGAAAAAAAAACACTGCAACAAGAGGGTA         CD5            72.87   \n",
       "1           1  TTTTAAAAAACCTACCGTAAACTCGGGTCA         NF1            65.80   \n",
       "2           2  TCAGAAAAAGCAGCGTCAGTGGATTGGCCC         CD5            84.21   \n",
       "3           3  AATAAAAAATAGGATTCCCAGCTTTGGAAG         NF1            56.39   \n",
       "4           4  GATGAAAAATATGTAAACAGCATTTGGGAC        CUL3             4.30   \n",
       "\n",
       "   Amino Acid Cut position  score_drug_gene_rank  score_drug_gene_threshold  \\\n",
       "0                    360.0              0.083682                          0   \n",
       "1                   1868.0              0.868207                          1   \n",
       "2                    416.0              0.184100                          0   \n",
       "3                   1601.0              0.432065                          0   \n",
       "4                     33.0              0.149351                          0   \n",
       "\n",
       "      drug  predictions  \n",
       "0   nodrug     0.544412  \n",
       "1  PLX_2uM     0.617512  \n",
       "2   nodrug     0.476232  \n",
       "3  PLX_2uM     0.459882  \n",
       "4  PLX_2uM     0.290841  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_fusi=pd.read_csv(\"Fusi_Data.csv\")\n",
    "df_fusi.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Need to find more data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Isolating necessary data into one dataframe\n",
    "- combine only necessary columns into one succinct dataframe\n",
    "    - include anything that could be useful to models\n",
    "- plot some graphs to visualize data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ian/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n",
      "/Users/ian/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Spacer Sequence</th>\n",
       "      <th>Expanded Sequence</th>\n",
       "      <th>sgRNA Score</th>\n",
       "      <th>sgRNA Rank</th>\n",
       "      <th>sgRNA Normalized</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTAATGGCTTCCTCGTGAGT</td>\n",
       "      <td>GGCCGTAATGGCTTCCTCGTGAGTTGGTCC</td>\n",
       "      <td>0.325738</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>0.789849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTCGTGCGAGTGATAGACC</td>\n",
       "      <td>AGAAGCTCGTGCGAGTGATAGACCAGGAGA</td>\n",
       "      <td>0.162559</td>\n",
       "      <td>707.0</td>\n",
       "      <td>0.552807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CACGGACGGTATATGGCAGG</td>\n",
       "      <td>ATGGCACGGACGGTATATGGCAGGAGGTGG</td>\n",
       "      <td>0.035653</td>\n",
       "      <td>191.0</td>\n",
       "      <td>0.149130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACTTCTGCGAGACACGGAC</td>\n",
       "      <td>TGGCTACTTCTGCGAGACACGGACAGGTCG</td>\n",
       "      <td>0.311042</td>\n",
       "      <td>982.0</td>\n",
       "      <td>0.767944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGGCCCACGCTACTACAGTT</td>\n",
       "      <td>ACTTCGGCCCACGCTACTACAGTTCGGGCT</td>\n",
       "      <td>0.199951</td>\n",
       "      <td>803.0</td>\n",
       "      <td>0.627909</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Spacer Sequence               Expanded Sequence  sgRNA Score  \\\n",
       "0  GTAATGGCTTCCTCGTGAGT  GGCCGTAATGGCTTCCTCGTGAGTTGGTCC     0.325738   \n",
       "1  GCTCGTGCGAGTGATAGACC  AGAAGCTCGTGCGAGTGATAGACCAGGAGA     0.162559   \n",
       "2  CACGGACGGTATATGGCAGG  ATGGCACGGACGGTATATGGCAGGAGGTGG     0.035653   \n",
       "3  TACTTCTGCGAGACACGGAC  TGGCTACTTCTGCGAGACACGGACAGGTCG     0.311042   \n",
       "4  CGGCCCACGCTACTACAGTT  ACTTCGGCCCACGCTACTACAGTTCGGGCT     0.199951   \n",
       "\n",
       "   sgRNA Rank  sgRNA Normalized  \n",
       "0      1010.0          0.789849  \n",
       "1       707.0          0.552807  \n",
       "2       191.0          0.149130  \n",
       "3       982.0          0.767944  \n",
       "4       803.0          0.627909  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## to do\n",
    "df_features=df_doench[[\"Spacer Sequence\",\"Expanded Sequence\",\"sgRNA Score\"]]\n",
    "df_features[\"sgRNA Rank\"] = df_features[\"sgRNA Score\"].rank(method='max')\n",
    "n = df_features[\"sgRNA Rank\"].size\n",
    "# D=(r − 3/8) / (n + 1/4)\n",
    "df_features[\"sgRNA Normalized\"] = (df_features[\"sgRNA Rank\"] - (3/8)) / (n + 1/4)\n",
    "df_features.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Normalization Verification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(df_V2[\"sgRNA Normalized\"], alpha=0.5, density=True, label='Normalized sgRNA scores')\n",
    "plt.hist(df_V2[\"sgRNA Score\"], alpha=0.5, density=True, label='Raw sgRNA scores')\n",
    "plt.ylabel('Probability density')\n",
    "plt.title(\"V2 Dataset Histogram sgRNA scores\")\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### V1 Featurization\n",
    "can definitely be improved upon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_V1 = df_features.copy(deep=True) ## creates a hard copy of supplementary table 7 data to work with\n",
    "df_V1['Sequence'] = df_V1['Sequence'].astype('string') ## sets data of sequence to String for parsing\n",
    "df_V1['Expanded Sequence'] = df_V1['Expanded Sequence'].astype('string') ## sets data of sequence to String for parsing\n",
    "df_V1['Expanded Sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "V1_bool_parameters=[\"GC<10\",\"GC>10\"]\n",
    "\n",
    "for k in range(30):\n",
    "    ## creates columns for each nucleotide at each position in sequence\n",
    "    for x in ['A',\"G\",\"C\",\"T\"]:\n",
    "        df_V1[x + str(k+1)] = letter_at_position(df_V1['Sequence'],x,k)\n",
    "        V1_bool_parameters.append(x + str(k+1))\n",
    "        if k < 29:\n",
    "            for y in [\"A\",\"G\",\"C\",\"T\"]:\n",
    "                df_V1[x + y + str(k+1)] = tuple_at_position(df_V1['Sequence'],x+y,k)\n",
    "                V1_bool_parameters.append(x + y + str(k+1))\n",
    "        else:\n",
    "            for y in [\"A\",\"G\",\"C\",\"T\"]:\n",
    "                df_V1[x + y + \"?\"] =  False # not technically correct but will work as placeholder\n",
    "                df_V1[x + y + \"_PAM\"] = False # this as well\n",
    "                V1_bool_parameters.append(x + y + \"?\")\n",
    "                V1_bool_parameters.append(x + y + \"_PAM\")\n",
    "        df_V1[x + \"?\"] = False # same with this\n",
    "        V1_bool_parameters.append(x + \"?\")\n",
    "        \n",
    "df_V1['GC Count'] = -1 ## creates column for GC count\n",
    "df_V1[\"PAM position (expanded sequence)\"] = -1\n",
    "\n",
    "df_V1['Melting Temp Wallace'] = 0\n",
    "df_V1['Melting Temp GC'] = 0\n",
    "df_V1['Melting Temp NN1'] = 0\n",
    "df_V1['Melting Temp NN2'] = 0\n",
    "df_V1['Melting Temp NN3'] = 0\n",
    "df_V1['Melting Temp NN4'] = 0\n",
    "\n",
    "df_V1[\"GC<10\"] = False\n",
    "df_V1[\"GC>10\"] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in range(df_V1['Sequence'].size):\n",
    "    ## loops through each sequence and computes the necessary one-hot nucleotide encoding features\n",
    "    for k in range(30):\n",
    "        for x in ['A',\"G\",\"C\",\"T\"]:\n",
    "            df_V1[x + str(k+1)][a] = letter_at_position(df_V1['Sequence'][a],x,k)\n",
    "            if k < 29:\n",
    "                for y in [\"A\",\"G\",\"C\",\"T\"]:\n",
    "                    df_V1[x + y + str(k+1)][a] = tuple_at_position(df_V1['Sequence'][a],x+y,k)\n",
    "                    if (not df_V1[x + y + \"?\"][a]):\n",
    "                        df_V1[x + y + \"?\"][a] = df_V1[x + y + str(k+1)][a]\n",
    "            \n",
    "            if (not df_V1[x + \"?\"][a]):\n",
    "                df_V1[x + \"?\"][a] = df_V1[x + str(k+1)][a]\n",
    "                        \n",
    "    ## generates the GC count data for each column\n",
    "    df_V1['GC Count'][a] = gc_count(df_V1['Sequence'][a])\n",
    "    ## calculates the PAM position for each sequence\n",
    "    df_V1[\"PAM position (expanded sequence)\"][a] = find_PAM(df_V1[\"Expanded Sequence\"][a],\"GG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in range(df_V1['Expanded Sequence'].size):\n",
    "    df_V1[\"PAM position (expanded sequence)\"][a] = find_PAM(df_V1[\"Expanded Sequence\"][a],\"GG\")\n",
    "    for x in ['A',\"G\",\"C\",\"T\"]:\n",
    "        for y in ['A',\"G\",\"C\",\"T\"]:\n",
    "            df_V1[x + y + \"_PAM\"] = tuple_at_position(df_V1[\"Expanded Sequence\"][a],\n",
    "                                                      x+y,\n",
    "                                                      df_V1[\"PAM position (expanded sequence)\"][a]-3)\n",
    "    seq = Seq(df_V1[\"Expanded Sequence\"][a])\n",
    "    df_V1[\"Melting Temp Wallace\"][a] = mt.Tm_Wallace(seq)\n",
    "    df_V1[\"Melting Temp GC\"][a] = mt.Tm_GC(seq)\n",
    "    df_V1[\"Melting Temp NN1\"][a] = mt.Tm_NN(seq, nn_table=mt.DNA_NN1)\n",
    "    df_V1[\"Melting Temp NN2\"][a] = mt.Tm_NN(seq, nn_table=mt.DNA_NN2)\n",
    "    df_V1[\"Melting Temp NN3\"][a] = mt.Tm_NN(seq, nn_table=mt.DNA_NN3)\n",
    "    df_V1[\"Melting Temp NN4\"][a] = mt.Tm_NN(seq, nn_table=mt.DNA_NN4)\n",
    "    \n",
    "    df_V1[\"GC<10\"][a] = df_V1[\"GC Count\"][a] < 10\n",
    "    df_V1[\"GC>10\"][a] = df_V1[\"GC Count\"][a] > 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for v1 in V1_bool_parameters:\n",
    "    df_V1[v1]=df_V1[v1].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_V1.to_csv('Doench2014_V1_Featurized.csv')\n",
    "df_V1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### V2 Featurization\n",
    "can definitely be improved upon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_V2 = df_V1.copy(deep=True)\n",
    "V2_bool_parameters=V1_bool_parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_V2[\"AAAAA\"] = \"AAAAA\" in df_V2[\"Expanded Sequence\"] ## initializes each column to false\n",
    "df_V2[\"GGGGG\"] = \"GGGGG\" in df_V2[\"Expanded Sequence\"]\n",
    "df_V2[\"CCCCC\"] = \"CCCCC\" in df_V2[\"Expanded Sequence\"]\n",
    "df_V2[\"TTTTT\"] = \"TTTTT\" in df_V2[\"Expanded Sequence\"]\n",
    "V2_bool_parameters.append([\"AAAAA\",\"GGGGG\",\"CCCCC\",\"TTTTT\"])\n",
    "df_V2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in range(29):\n",
    "    for x in ['A',\"G\",\"C\",\"T\"]:\n",
    "        for y in ['A',\"G\",\"C\",\"T\"]:\n",
    "            for z in [\"A\",\"G\",\"C\",\"T\"]:\n",
    "                \n",
    "                if k < 28:\n",
    "                    df_V2[x + y + z + str(k+1)] = False\n",
    "                    V2_bool_parameters.append(x + y + z + str(k+1))\n",
    "                else:\n",
    "                    df_V2[x+y+z] = False\n",
    "                    V2_bool_parameters.append(x+y+z)\n",
    "df_V2[\"Delta G\"] = 0\n",
    "df_V2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in range(df_V2['Expanded Sequence'].size):\n",
    "    df_V2[\"AAAAA\"][a] = \"AAAAA\" in df_V2[\"Expanded Sequence\"][a]\n",
    "    df_V2[\"GGGGG\"][a] = \"GGGGG\" in df_V2[\"Expanded Sequence\"][a]\n",
    "    df_V2[\"CCCCC\"][a] = \"CCCCC\" in df_V2[\"Expanded Sequence\"][a]\n",
    "    df_V2[\"TTTTT\"][a] = \"TTTTT\" in df_V2[\"Expanded Sequence\"][a]\n",
    "    \n",
    "    for k in range(29):\n",
    "        for x in ['A',\"G\",\"C\",\"T\"]:\n",
    "            for y in ['A',\"G\",\"C\",\"T\"]:\n",
    "                for z in [\"A\",\"G\",\"C\",\"T\"]:\n",
    "\n",
    "                    if k < 28:\n",
    "                        df_V2[x + y + z + str(k+1)][a] = triple_at_position(df_V2[\"Expanded Sequence\"][a], x+y+z, k)\n",
    "                        if (not df_V2[x+y+z][a]):\n",
    "                            df_V2[x+y+z][a] = df_V2[x + y + z + str(k+1)][a]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in range(df_V2['Sequence'].size):\n",
    "    ## note that this uses the 20-n.t. guide RNA sequence\n",
    "    ## all previously calculated features use the 30-oligo target sequence\n",
    "    \n",
    "    # # Calculate melting temperature for 25uM oligos\n",
    "    # (name, g, h, s, tm, seq)\n",
    "    df_V2[\"Delta G\"][a] = OligoMelt.Duplex.calc_tm(df_V2[\"Sequence\"][a])[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for v2 in V2_bool_parameters:\n",
    "    df_V2[v2]=df_V2[v2].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_V2.to_csv('Doench2014_V2_Featurized.csv')\n",
    "df_V2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### V3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# to do"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "9d95a0750442c48154ac88d02c074999fe7569ac5db8196fc6e24e1e7ebc41bc"
  },
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
